HW-DAO9

Max-Pochepets opened the pull request:
Authentication and Authorization implemented

Max-Pochepets committed:
Initial

Max-Pochepets committed:
checklist demands fulfilled

Max-Pochepets committed:
Fix: findByLogin query improved

Mandybur800 commented on src/main/java/crud/model/Driver.java:
```diff
@@ -6,10 +6,14 @@
     private String licenseNumber;
+    private String login;
+    private String password;
 
-    public Driver(String name, String licenceNumber) {
+    public Driver(String name, String licenceNumber, String login, String password) {
```
Better to add a new constructor, because you can use the old constructor in other places

Mandybur800 commented on src/main/java/crud/web/filters/AuthenticationFilter.java:
```diff
@@ -0,0 +1,46 @@
+            return;
+        }
+        Long driverId = (Long) req.getSession().getAttribute(DRIVER_ID);
+        if (driverId == null || driverService.get(driverId) == null) {
```
Please, check the 5th point in [common mistakes](https://mate-academy.github.io/jv-program-fulltime/02_jdbc/checklist/09_filters_checklist.html)

Max-Pochepets committed:
Fix: redundant check in Filter removed

Max-Pochepets committed:
Fix:

minor changes to handle a situation if a logged in user deletes themselves

RomanDubovskyi commented on src/main/java/crud/controller/IndexController.java:
```diff
@@ -9,6 +11,10 @@
 
 public class IndexController extends HttpServlet {
+    private static final Injector INJECTOR = Injector.getInstance("crud");
+    private final DriverService driverService
```
do we use it here?

RomanDubovskyi commented on src/main/java/crud/controller/security/AuthenticationServiceImpl.java:
```diff
@@ -0,0 +1,23 @@
+@ServiceImpl
+public class AuthenticationServiceImpl implements AuthenticationService {
+    @Inject
+    DriverService driverService;
```
what about access modifiers?

RomanDubovskyi commented on src/main/java/crud/web/filters/AuthenticationFilter.java:
```diff
@@ -0,0 +1,46 @@
+            return;
+        }
+        Long driverId = (Long) req.getSession().getAttribute(DRIVER_ID);
+        if (driverId == null || driverService.get(driverId) == null) {
```
does your service return null or it throws an exception?

Max-Pochepets replied:
driver service returns null. I know, that returning null is a bad practice, but it was made to handle a situation where logged in user deletes themselves and for this user not to see an exception stacktrace in their browser.

> does your service return null or it throws an exception?

RomanDubovskyi replied:
that makes program structure a little confusing (why do we need to return optional from dao to return null from service if there is not object). Let's not change the service layer and not handle this particular case. In real app you will use Spring security and it's possible to invalidate a users' session when you delete one so it won't be a problem.

RomanDubovskyi reviewed (changes requested):
don't forget about the points, mentioned in the chat

Max-Pochepets committed:
Fix: minor changes

code duplication problem in AuthenticationServiceImpl resolved
accessible for guest user urls now stored in Set
access modifiers added
redundant variable removed

RomanDubovskyi commented on src/main/java/crud/web/filters/AuthenticationFilter.java:
```diff
@@ -0,0 +1,51 @@
+
+public class AuthenticationFilter implements Filter {
+    private static final String DRIVER_ID = "driver_id";
+    private static final Injector INJECTOR = Injector.getInstance("crud");
```
```suggestion
    private static final Injector injector = Injector.getInstance("crud");
```
it's not a constant because the object is not deeply immutable

RomanDubovskyi commented on src/main/java/crud/web/filters/AuthenticationFilter.java:
```diff
@@ -0,0 +1,51 @@
+public class AuthenticationFilter implements Filter {
+    private static final String DRIVER_ID = "driver_id";
+    private static final Injector INJECTOR = Injector.getInstance("crud");
+    private final Set<String> urls = new HashSet<>();
```
```suggestion
    private final Set<String> allowedUrls = new HashSet<>();
```

Max-Pochepets committed:
Refactor:

naming fixed
DriverServiceImpl logic reverted

RomanDubovskyi commented on src/main/java/crud/web/filters/AuthenticationFilter.java:
```diff
@@ -0,0 +1,51 @@
+
+public class AuthenticationFilter implements Filter {
+    private static final String DRIVER_ID = "driver_id";
+    private static final Injector injector = Injector.getInstance("crud");
+    private final Set<String> allowedUrls = new HashSet<>();
+    private final DriverService driverService
+            = (DriverService) injector.getInstance(DriverService.class);
+
```
```suggestion
    private final Set<String> allowedUrls = new HashSet<>();
```
we don't use it anymore

RomanDubovskyi reviewed (approved):
remove redundant fields from filter, rest LGTM

Max-Pochepets committed:
Refactor:

redundant fields from AuthenticationFilter removed
